File: services/frontend/src/stores/activity.ts
import { defineStore } from 'pinia'
import { ref } from 'vue'
import { activityApi } from '@/api/activity'
import type { ActivityEvent, ActivitySummary } from '@/api/types'

export const useActivityStore = defineStore('activity', () => {
  const events = ref<ActivityEvent[]>([])
  const total = ref(0)
  const summary = ref<ActivitySummary | null>(null)
  const eventTypeCounts = ref<Record<string, number>>({})
  const loading = ref(false)
  const error = ref<string | null>(null)

  async function fetchFeed(params?: {
    limit?: number
    offset?: number
    event_type?: string
    severity?: string
    movie_id?: number
  }) {
    loading.value = true
    error.value = null
    try {
      const response = await activityApi.getFeed(params)
      events.value = response.events
      total.value = response.total
    } catch (e: any) {
      error.value = e.message || 'Failed to fetch activity feed'
    } finally {
      loading.value = false
    }
  }

  async function fetchSummary(hours?: number) {
    try {
      summary.value = await activityApi.getSummary(hours)
    } catch (e: any) {
      error.value = e.message
    }
  }

  async function fetchEventTypeCounts() {
    try {
      eventTypeCounts.value = await activityApi.getEventTypeCounts()
    } catch (e: any) {
      error.value = e.message
    }
  }

  return {
    events,
    total,
    summary,
    eventTypeCounts,
    loading,
    error,
    fetchFeed,
    fetchSummary,
    fetchEventTypeCounts,
  }
})

File: services/api/app/api/v1/activity.py
"""
Activity Feed API Endpoints
Chronological timeline of all significant events
"""
from typing import Any

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from app.core.database import get_db
from app.core.logging import get_logger
from app.services.activity_service import ActivityService

router = APIRouter()
logger = get_logger(__name__)


@router.get("")
@router.get("/", include_in_schema=False)
async def get_activity_feed(
    limit: int = Query(50, ge=1, le=200, description="Items per page"),
    offset: int = Query(0, ge=0, description="Offset"),
    event_type: str | None = Query(None, description="Filter by event type"),
    severity: str | None = Query(None, description="Filter by severity"),
    movie_id: int | None = Query(None, description="Filter by movie ID"),
    db: AsyncSession = Depends(get_db),
):
    """
    Get the activity feed with filtering and pagination.

    Event types: movie_added, movie_upgraded, movie_removed,
    download_approved, download_declined, scan_completed,
    collection_changed, upgrade_available, ipt_scan
    """
    service = ActivityService(db)
    events, total = await service.get_feed(
        limit=limit,
        offset=offset,
        event_type=event_type,
        severity=severity,
        movie_id=movie_id,
    )

    return {
        "total": total,
        "limit": limit,
        "offset": offset,
        "events": [
            {
                "id": e.id,
                "event_type": e.event_type,
                "title": e.title,
                "description": e.description,
                "severity": e.severity,
                "movie_id": e.movie_id,
                "movie_title": e.movie_title,
                "movie_year": e.movie_year,
                "related_id": e.related_id,
                "related_type": e.related_type,
                "quality_before": e.quality_before,
                "quality_after": e.quality_after,
                "metadata": e.extra_data,
                "created_at": e.created_at.isoformat() if e.created_at else None,
            }
            for e in events
        ],
    }


@router.get("/summary")
async def get_activity_summary(
    hours: int = Query(24, ge=1, le=720, description="Hours to look back"),
    db: AsyncSession = Depends(get_db),
):
    """Get summary of recent activity"""
    service = ActivityService(db)
    return await service.get_recent_summary(hours)


@router.get("/types")
async def get_event_type_counts(db: AsyncSession = Depends(get_db)):
    """Get count of events by type"""
    service = ActivityService(db)
    return await service.get_event_type_counts()


@router.get("/movie/{movie_id}")
async def get_movie_timeline(
    movie_id: int,
    db: AsyncSession = Depends(get_db),
):
    """Get all activity events for a specific movie"""
    service = ActivityService(db)
    events = await service.get_movie_timeline(movie_id)

    return [
        {
            "id": e.id,
            "event_type": e.event_type,
            "title": e.title,
            "description": e.description,
            "severity": e.severity,
            "quality_before": e.quality_before,
            "quality_after": e.quality_after,
            "metadata": e.extra_data,
            "created_at": e.created_at.isoformat() if e.created_at else None,
        }
        for e in events
    ]
